option cc_generic_services = true;


message ShardBeginArgs {
  required int32 Tid = 1;  // 事务ID
};

message ShardBeginReply {
  required int32 err = 1; // RPC操作
};

message ShardEndArgs {
  required int32 Tid = 1;
  required int32 commitID = 2;
};

message ShardEndReply {
  required int32 err = 1;
};

message ShardAbortArgs {
  required int32 Tid = 1;
};

message ShardAbortReply {
  required int32 err = 1;
};


message ShardReadArgs {
  required int32 Tid = 1;
  required string key = 2;
};

message ShardReadReply {
  required int32 err = 1;
  optional string value = 2; // optional
};

message ShardWriteArgs {
  required int32 Tid = 1;
  required string key = 2;
  required string value = 3;

  // 幂等性先不考虑
  optional int64 clientID = 4;
  optional int64 seqID = 5;
};

message ShardWriteReply {
  required int32 err = 1;
};

// coordinator
message ConnectArgs {
  required string host = 1;
};

message ConnectReply {
  required int32 err = 1;
};


service ShardKvBase {
  rpc PrepareRead(ShardReadArgs) returns (ShardReadReply);
  rpc PrepareWrite(ShardWriteArgs) returns (ShardWriteReply);

  rpc BEGIN(ShardBeginArgs) returns (ShardBeginReply);
  rpc END(ShardEndArgs) returns (ShardEndReply);
  rpc ABORT(ShardAbortArgs) returns (ShardAbortReply);

  rpc CONNECT(ConnectArgs) returns (ConnectReply);
}





